controller/script/simulate-proxy/main.go

```diff
@@ -244,109 +360,69 @@ func main() {
+		if pod.Status.PodIP != "" && !strings.HasPrefix(pod.GetNamespace(), "kube-") && (*maxPods == 0 || len(allPods) < *maxPods) {
 			allPods = append(allPods, pod)
+			podsByNamespace[pod.GetNamespace()] = append(podsByNamespace[pod.GetNamespace()], pod)
 		}
 	}
 
-	for i := 0; (*requestCount == 0) || (i < *requestCount); i++ {
-		count := randomCount()
-		sourceIp := randomPod(allPods, nil)
-		targetIp := randomPod(allPods, sourceIp)
-
-		req := &pb.ReportRequest{
-			Process: &pb.Process{
-				ScheduledInstance:  "hello-1mfa0",
-				ScheduledNamespace: "people",
-			},
-			ClientTransports: []*pb.ClientTransport{
-				// TCP
-				&pb.ClientTransport{
-					TargetAddr: &common.TcpAddress{
-						Ip:   targetIp,
-						Port: randomPort(),
-					},
-					Connects: count,
-					Disconnects: []*pb.TransportSummary{
-						&pb.TransportSummary{
-							DurationMs: uint64(randomCount()),
-							BytesSent:  uint64(randomCount()),
-						},
-					},
-					Protocol: common.Protocol_TCP,
-				},
-			},
-			ServerTransports: []*pb.ServerTransport{
-				// TCP
-				&pb.ServerTransport{
-					SourceIp: sourceIp,
-					Connects: count,
-					Disconnects: []*pb.TransportSummary{
-						&pb.TransportSummary{
-							DurationMs: uint64(randomCount()),
-							BytesSent:  uint64(randomCount()),
-						},
-					},
-					Protocol: common.Protocol_TCP,
-				},
-			},
-			Proxy: pb.ReportRequest_INBOUND,
-			Requests: []*pb.RequestScope{
-
-				// gRPC
-				&pb.RequestScope{
-					Ctx: &pb.RequestCtx{
-						SourceIp: sourceIp,
-						TargetAddr: &common.TcpAddress{
-							Ip:   targetIp,
-							Port: randomPort(),
-						},
-						Authority: "world.greeting:7778",
-					},
-					Count: count,
-					Responses: []*pb.ResponseScope{
-						&pb.ResponseScope{
-							Ctx: &pb.ResponseCtx{
-								HttpStatusCode: http.StatusOK,
-							},
-							ResponseLatencyCounts: randomLatencies(count),
-							Ends: randomGrpcEos(count),
-						},
-					},
-				},
-
-				// HTTP/2
-				&pb.RequestScope{
-					Ctx: &pb.RequestCtx{
-						SourceIp: sourceIp,
-						TargetAddr: &common.TcpAddress{
-							Ip:   targetIp,
-							Port: randomPort(),
-						},
-						Authority: "world.greeting:7778",
-					},
-					Count: count,
-					Responses: []*pb.ResponseScope{
-						&pb.ResponseScope{
-							Ctx: &pb.ResponseCtx{
-								HttpStatusCode: randomHttpResponseCode(),
-							},
-							ResponseLatencyCounts: randomLatencies(count),
-							Ends: randomH2Eos(count),
-						},
-					},
-				},
-			},
-
-			HistogramBucketBoundsTenthMs: latencyBucketBounds[:],
-		}
-
-		_, err = client.Report(context.Background(), req)
-		if err != nil {
-			log.Fatal(err.Error())
-		}
-
-		time.Sleep(*sleep)
+	stopCh := make(chan os.Signal)
+	signal.Notify(stopCh, os.Interrupt, os.Kill)
+
+	for _, addr := range strings.Split(*metricsAddrs, ",") {
+		randomNamespace := getRandomNamespaceKey(podsByNamespace)
+		proxyPod := podList[rand.Intn(len(podList))]
```

Comment:
One disadvantage of doing it this way is that the same pod could be selected more than once, which means you'll have multiple reporters simulating the same pod

Reply:
Hmm, you are right. We could remove the key value pair when it is returned from `getRandomNamespaceKey` and is used by the generated proxy. If we end up removing all namespaces from the `podsByNamespace` value, we can start to generate random namespaces i.e. like a `uuid` string.

Reply:
yeah.  I think this plays nicely with the model where we don't get data from a real k8s cluster but instead generate a configurable number of random synthetic pods and serve metrics for all of those pods.